helix-gateway/src/main/java/org/apache/helix/gateway/util/GatewayCurrentStateCache.java

```diff
@@ -0,0 +1,182 @@
+  static ObjectMapper mapper = new ObjectMapper();
+  String _clusterName;
+
+  // A cache of current state. It should be updated by the HelixGatewayServiceChannel
+  // instance -> resource state (resource -> shard -> target state)
+  Map<String, ShardStateMap> _currentStateMap;
+
+  // A cache of target state.
+  // instance -> resource state (resource -> shard -> target state)
+  Map<String, ShardStateMap> _targetStateMap;
+
+  public GatewayCurrentStateCache(String clusterName) {
+    _clusterName = clusterName;
+    _currentStateMap = new HashMap<>();
+    _targetStateMap = new HashMap<>();
+  }
+
+  public String getCurrentState(String instance, String resource, String shard) {
+    return _currentStateMap.get(instance).getState(resource, shard);
+  }
+
+  public String getTargetState(String instance, String resource, String shard) {
+    return _targetStateMap.get(instance).getState(resource, shard);
+  }
+
+  /**
+   * Update the cached current state of instances in a cluster, and return the diff of the change.
+   * @param newCurrentStateMap The new current state map of instances in the cluster
+   * @return
+   */
+  public Map<String, Map<String, Map<String, String>>> updateCacheWithNewCurrentStateAndGetDiff(
+      Map<String, Map<String, Map<String, String>>> newCurrentStateMap) {
+    Map<String, Map<String, Map<String, String>>> diff = new HashMap<>();
+    for (String instance : newCurrentStateMap.keySet()) {
+      Map<String, Map<String, String>> newCurrentState = newCurrentStateMap.get(instance);
+      diff.put(instance, _currentStateMap.computeIfAbsent(instance, k -> new ShardStateMap(new HashMap<>()))
+          .updateMapAndGetDiff(newCurrentState));
+    }
+    return diff;
+  }
+
+  /**
+   * Update the cache with the current state diff.
+   * All existing target states remains the same
+   * @param currentStateDiff
+   */
+  public void updateCacheWithCurrentStateDiff(Map<String, Map<String, Map<String, String>>> currentStateDiff) {
+    for (String instance : currentStateDiff.keySet()) {
+      Map<String, Map<String, String>> currentStateDiffMap = currentStateDiff.get(instance);
+      _currentStateMap.computeIfAbsent(instance, k -> new ShardStateMap(new HashMap<>()))
+          .updateShardStateMapWithDiff(currentStateDiffMap);
+    }
+  }
+
+  /**
+   * Update the target state with the changed target state maps.
+   * All existing target states remains the same
+   * @param targetStateChangeMap
+   */
+  public void updateTargetStateWithDiff(String instance, Map<String, Map<String, String>> targetStateChangeMap) {
+    _targetStateMap.computeIfAbsent(instance, k -> new ShardStateMap(new HashMap<>()))
+        .updateShardStateMapWithDiff(targetStateChangeMap);
+  }
+
+  /**
+   * Serialize the target state assignments to a JSON Node.
+   * example ： {"instance1":{"resource1":{"shard1":"ONLINE","shard2":"OFFLINE"}}}}
+   */
+  public ObjectNode serializeTargetAssignmentsToJSON() {
```

Comment:
What about timestamp?

Reply:
This is cluster level info. Timestamp will be added when we aggregate all clusters info and write to file.